src/core/MIGRATION.md

```diff
@@ -0,0 +1,683 @@
+
+The basic file structure of a new platform plugin named "demo" that had both client-side and server-side code would be:
+
+```
+src/plugins
+  demo
+    kibana.json [1]
+    public
+      index.ts [2]
+      plugin.ts [3]
+    server
+      index.ts [4]
+      plugin.ts [5]
+```
+
+**[1] `kibana.json`** is a static manifest file that is used to identify the plugin and to determine what kind of code the platform should execute from the plugin:
+
+```json
+{
+  "id": "demo",
+  "server": true,
+  "ui": true
+}
+```
+
+Note that `package.json` files are irrelevant to and ignored by the new platform.
+
+**[2] `public/index.ts`** is the entry point into the client-side code of this plugin. It must export a function named `plugin`, which will receive a standard set of core capabilities as an argument (e.g. logger). It should return an instance of its plugin definition for the platform to register at load time.
+
+```ts
+import { PluginInitializerContext } from '../../../core/public';
+import { Plugin } from './plugin';
+
+export function plugin(initializerContext: PluginInitializerContext) {
+  return new Plugin(initializerContext);
+}
+```
+
+**[3] `public/plugin.ts`** is the client-side plugin definition itself. Technically speaking it does not need to be a class or even a separate file from the entry point, but _all plugins at Elastic_ should be consistent in this way.
+
+```ts
+import { PluginInitializerContext, CoreSetup, PluginStop } from '../../../core/public';
+
+export class Plugin {
+  constructor(initializerContext: PluginInitializerContext) {
+  }
+
+  public setup(core: CoreSetup) {
+    // called when plugin is setting up
+  }
+
+  public stop(core: PluginStop) {
+    // called when plugin is torn down, aka window.onbeforeunload
+  }
+}
+```
+
+**[4] `server/index.ts`** is the entry-point into the server-side code of this plugin. It is identical in almost every way to the client-side entry-point:
+
+```ts
+import { PluginInitializerContext } from '../../../core/server';
+import { Plugin } from './plugin';
+
+export function plugin(initializerContext: PluginInitializerContext) {
+  return new Plugin(initializerContext);
+}
+```
+
+**[5] `server/plugin.ts`** is the server-side plugin definition. The _shape_ of this plugin is the same as it's client-side counter-part:
+
+```ts
+import { PluginInitializerContext, CoreSetup, PluginStop } from '../../../core/server';
+
+export class Plugin {
+  constructor(initializerContext: PluginInitializerContext) {
+  }
+
+  public setup(core: CoreSetup) {
+    // called when plugin is setting up during Kibana's startup sequence
+  }
+
+  public stop(core: PluginStop) {
```

Comment:
Perhaps `PluginStop` should be called `CoreStop` to make it consistent with `CoreStart`, `CoreSetup`